pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,449 @@
+		ac.keepAlive.keepAlive.Stop()
+		ac.keepAlive = nil
+	}
+
+	ac.keepAlive = &keepAliveListener{
+		keepAlive: hb,
+		conn:      ac.host.Connection,
+	}
+
+	// this should be equal to
+	// ac.codecClient.AddConnectionEventListener(ac.keepAlive)
+	ac.host.Connection.AddConnectionEventListener(ac.keepAlive)
+}
+
+// keepAliveListener is a types.ConnectionEventListener
+type keepAliveListener struct {
+	keepAlive KeepAlive
+	conn      api.Connection
+}
+
+// OnEvent impl types.ConnectionEventListener
+func (l *keepAliveListener) OnEvent(event api.ConnectionEvent) {
+	if event == api.OnReadTimeout && l.keepAlive != nil {
+		heartbeatFailCreator := func() {
+			l.conn.Close(api.NoFlush, api.LocalClose)
+		}
+
+		// TODO, whether error should be handled
+		l.conn.Write(buffer.NewIoBufferBytes(l.keepAlive.GetKeepAliveData(heartbeatFailCreator)))
+	}
+}
+
+type KeepAlive interface {
+	Stop()
+
+	GetKeepAliveData(failCallback func()) []byte
+}
+
+type Connection interface {
+	Write(buf ...buffer.IoBuffer) error
+	Destroy()
+	Available() bool
+}
+
+// NewConn returns a simplified connpool
+func NewConn(hostAddr string, reconnectTryTimes int, heartBeatCreator func() KeepAlive, readFilters []api.ReadFilter, autoReconnectWhenClose bool) Connection {
+	// use host addr as cluster name, for the count of metrics
+	cl := basicCluster(hostAddr, []string{hostAddr})
+	host := cluster.NewSimpleHost(cl.Hosts[0], cluster.NewCluster(cl).Snapshot().ClusterInfo())
+
+	// if user configure this to -1, then retry is unlimited
+	if reconnectTryTimes == -1 {
+		reconnectTryTimes = math.MaxInt32
+	}
+
+	p := &connpool{
+		idleClients: make([]*activeClient, 0, 1),
+		host:        host,
+
+		heartBeatCreator:       heartBeatCreator,
+		autoReconnectWhenClose: autoReconnectWhenClose,
+		reconnTryTimes:         reconnectTryTimes,
+		readFilters:            readFilters,
+	}
+
+	// trigger the client generation
+	p.GetActiveClient(context.Background())
+
+	return p
+}
+
+// write to client
+func (p *connpool) Write(buf ...buffer.IoBuffer) error {
+	cli, reason := p.GetActiveClient(context.Background())
+	if reason != "" {
+		return errors.New("get client failed" + string(reason))
+	}
+
+	// use atomic to load pointer, avoid partial pointer read
+	h := (*types.CreateConnectionData)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&cli.host))))
+
+	err := h.Connection.Write(buf...)
+	if err != nil {
+		cli.Reconnect()
+	}
+
+	return err
+}
+
+// Available current available to send request
+// WARNING, this api is only for msg
+// WARNING, dont use it any other scene
+func (p *connpool) Available() bool {
+	// if pool was destroyed
+	if atomic.LoadUint64(&p.destroyed) == 1 {
+		return false
+	}
+
+	// if there is no client
+	lastIdx := len(p.idleClients) - 1
+	if lastIdx < 0 {
+		return false
+	}
+
+	// if we are reconnecting
+	if atomic.LoadUint64(&p.idleClients[lastIdx].reconnectState) == connecting {
+		return false
+	}
+
+	return true
```

Comment:
Why not determine whether the connection is  writable?

Reply:
it's hard to know the connection is writable until you write to it...